### PR TITLE
Improvements to vsync abstraction

### LINKED ISSUE #839: Fix vsync issue when document is not visible.
-- better Title and desc to be provided for this bug.

### LINKED ISSUE #841: Add simple polyfill for requestAnimationFrame
This gets us a lot of browser compat for very little work.

### PR DESCRIPTION
**Pull request by cramforce**
- batching of measure/mutate in vsync.
- non-execution of vsync when doc is (virtually) invisible.
- polyfill for RAF
- tests for all that stuff.

Fixes #841
Fixes #839

### CODE REVIEW

**Comment by cramforce** [comment]
CC @sriramkrish85

File: src/vsync.js [hunk 1]
```diff
@@ -43,7 +44,26 @@ export class Vsync {
...
+    this.tasks_ = [];
+
+    /**
+     * States for tasks in the next frame in the same order.
+     * @private {!Array<!Object>}
+     */
+    this.states_ = [];
+
+    /**
+     * Whether a new animation frame has been scheduled.
```
Comment (dvoytenko) [inline src/vsync.js]: @private {...}

File: src/vsync.js [hunk 2]
```diff
@@ -51,12 +71,36 @@ export class Vsync {
    * @param {!Object<string, *>|undefined} opt_state
    */
   run(task, opt_state) {
+    // Do not request animation frames when the document is not visible.
+    if (!viewerFor(this.win).isVisible()) {
```
Comment (dvoytenko) [inline src/vsync.js]: What's better here? viewer.isVisible or docState.isVisible?

File: src/vsync.js [hunk 3]
```diff
@@ -51,12 +71,36 @@ export class Vsync {
...
+      return;
```
Comment (dvoytenko) [inline src/vsync.js]: Should we consider assert in this case? Or at least an API that returns `false` if scheduling wasn't possible?

File: src/vsync.js [hunk 4]
```diff
@@ -51,12 +71,36 @@ export class Vsync {
...
+      const states = this.states_;
+      this.tasks_ = [];
+      this.states_ = [];
+      for (let i = 0; i < tasks.length; i++) {
+        if (tasks[i].measure) {
+          tasks[i].measure(states[i]);
+        }
+      }
+      for (let i = 0; i < tasks.length; i++) {
+        if (tasks[i].mutate) {
```
Comment (dvoytenko) [inline src/vsync.js]: Could you please modify the definition of `mutate` above in `VsyncTask` struct as optional?

File: src/vsync.js [hunk 5]
```diff
@@ -112,6 +156,26 @@ export class Vsync {
...
   }
+
+  /**
+   * @return {function(function())} requestAnimationFrame or polyfill.
+   */
+  getRaf_() {
+    const raf = this.win.requestAnimationFrame
+        || this.win.webkitRequestAnimationFrame;
+    if (raf) {
+      return raf.bind(window);
```
Comment (erwinmombay) [inline src/vsync.js]: does it make sense to bind to `this.win` here for consistency?

Comment (cramforce) [inline src/vsync.js]: Yes, done.

File: src/vsync.js (hunk 4 above)
Comment (cramforce) [inline src/vsync.js]: Done

File: src/vsync.js (hunk 3 above)
Comment (cramforce) [inline src/vsync.js]: Added a warning.

File: src/vsync.js (hunk 2 above)
Comment (cramforce) [inline src/vsync.js]: viewer.isVisible is false when docState.isVisible or the iframe is hidden, right?

I think that is what we want.

File: src/vsync.js (hunk 1 above)
Comment (cramforce) [inline src/vsync.js]: Done

**Comment by cramforce** [comment]
Added docs explaining the new behavior.

File: src/vsync.js (hunk 2 above)
Comment (dvoytenko) [inline src/vsync.js]: Yes. Agree. Let's do it this way.

**Comment by dvoytenko** [comment]
LGTM

**Pull request merged**